notebook/static/notebook/js/codecell.js

```diff
@@ -335,10 +335,10 @@ define([
         this.render();
         this.events.trigger('execute.CodeCell', {cell: this});
         var that = this;
-        this.events.on('finished_iopub.Kernel', function (evt, data) {
+        this.events.one('finished_iopub.Kernel', function (evt, data) {
             if (that.kernel.id === data.kernel.id && that.last_msg_id === data.msg_id) {
```

Comment:
When `'finished_iopub.Kernel'` is triggered and its data matches that of the CodeCell, then we know that execution is complete.